ddtrace/internal/flare.py

```diff
@@ -0,0 +1,184 @@
+from ddtrace.internal.logger import get_logger
+from ddtrace.internal.utils.http import get_connection
+
+
+TRACER_FLARE_DIRECTORY = "tracer_flare"
+TRACER_FLARE_TAR = f"{TRACER_FLARE_DIRECTORY}.tar"
+TRACER_FLARE_ENDPOINT = "/tracer_flare/v1"
+TRACER_FLARE_FILE_HANDLER_NAME = "tracer_flare_file_handler"
+TRACER_FLARE_LOCK = "tracer_flare.lock"
+DEFAULT_TIMEOUT_SECONDS = 5
+
+log = get_logger(__name__)
+
+
+class Flare:
+    def __init__(self, timeout_sec: int = DEFAULT_TIMEOUT_SECONDS):
+        self.original_log_level = 0
+        self.timeout = timeout_sec
+
+    def _prepare(self, configs: List[dict]):
+        """
+        Update configurations to start sending tracer logs to a file
+        to be sent in a flare later.
+        """
+        os.makedirs(TRACER_FLARE_DIRECTORY, exist_ok=True)
+        for c in configs:
+            # AGENT_CONFIG is currently being used for multiple purposes
+            # We only want to prepare for a tracer flare if the config name
+            # starts with 'flare-log-level'
+            if not c.get("name", "").startswith("flare-log-level"):
+                return
+
+            # Validate the flare log level
+            flare_log_level = c.get("config", {}).get("log_level").upper()
+            flare_log_level_int = logging.getLevelName(flare_log_level)
+            if type(flare_log_level_int) != int:
+                raise TypeError("Invalid log level provided: %s", flare_log_level_int)
+
+            ddlogger = get_logger("ddtrace")
+            pid = os.getpid()
+            flare_file_path = f"{TRACER_FLARE_DIRECTORY}/tracer_python_{pid}.log"
+            self.original_log_level = ddlogger.level
+
+            # Set the logger level to the more verbose between original and flare
+            # We do this valid_original_level check because if the log level is NOTSET, the value is 0
+            # which is the minimum value. In this case, we just want to use the flare level, but still
+            # retain the original state as NOTSET/0
+            valid_original_level = 100 if self.original_log_level == 0 else self.original_log_level
+            logger_level = min(valid_original_level, flare_log_level_int)
+            ddlogger.setLevel(logger_level)
+            _add_file_handler(ddlogger, flare_file_path, flare_log_level, TRACER_FLARE_FILE_HANDLER_NAME)
+
+            # Create and add config file
+            self._generate_config_file(pid)
+
+    def _send(self, configs: List[Any]):
+        """
+        Revert tracer flare configurations back to original state
+        before sending the flare.
+        """
+        for c in configs:
+            # AGENT_TASK is currently being used for multiple purposes
+            # We only want to generate the tracer flare if the task_type is
+            # 'tracer_flare'
+            if type(c) != dict or c.get("task_type") != "tracer_flare":
+                continue
+            args = c.get("args", {})
+
+            self._revert_configs()
```

Comment:
All of this is happening inside of a loop. Is it your intention to do the revert once per `config`? This could be simplified by decoupling `Flare` from the config format.

Reply:
Ah, good catch. No this should just happen once. I had to add a loop because each RC product _could_ have multiple configs included, and only one of them is the tracer_flare request. Either way, we already have the expectation that we can only process one tracer flare request at a time. I'll add a return statement in the `finally` block so that we just process one and exit.

Reply:
I guess the one thing with having the return is that in the very odd case that there are two tracer flare configs, we wouldn't be logging the warning about skipping the second one, but I don't think it's a huge deal.